### ISSUE TITLE
DefaultAssemblyPartDiscoveryProvider throws KeyNtoFoundException if a library does not have runtime components

### ISSUE THREAD

**Issue by Gonkers** [aspnet/Mvc#5690]
Hello,
I realize what I'm doing isn't supported, but maybe someone can help. I am attempting to run an mvc core web app on my RaspberryPI 3 but I'm getting a KeyNotFoundException coming from the AddMvc() extension method.

For my current setup I'm using a build of dotnet core for RPi made by @ayende (https://ravendb.net/downloads). I am using VS2017 and a new mvc core app template (no changes). here is the stack trace
```
Unhandled Exception: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.Collections.Generic.KeyNotFoundException: The given key was not present in the dictionary.
   at System.ThrowHelper.ThrowKeyNotFoundException()
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at Microsoft.AspNetCore.Mvc.Internal.DefaultAssemblyPartDiscoveryProvider.CandidateResolver.ComputeClassification(String dependency)
   at Microsoft.AspNetCore.Mvc.Internal.DefaultAssemblyPartDiscoveryProvider.CandidateResolver.ComputeClassification(String dependency)
   at Microsoft.AspNetCore.Mvc.Internal.DefaultAssemblyPartDiscoveryProvider.CandidateResolver.ComputeClassification(String dependency)
   at Microsoft.AspNetCore.Mvc.Internal.DefaultAssemblyPartDiscoveryProvider.CandidateResolver.ComputeClassification(String dependency)
   at Microsoft.AspNetCore.Mvc.Internal.DefaultAssemblyPartDiscoveryProvider.CandidateResolver.<GetCandidates>d__4.MoveNext()
   at System.Linq.Enumerable.<SelectManyIterator>d__160`2.MoveNext()
   at System.Linq.Enumerable.SelectEnumerableIterator`2.MoveNext()
   at Microsoft.Extensions.DependencyInjection.MvcCoreServiceCollectionExtensions.GetApplicationPartManager(IServiceCollection services)
   at Microsoft.Extensions.DependencyInjection.MvcCoreServiceCollectionExtensions.AddMvcCore(IServiceCollection services)
   at Microsoft.Extensions.DependencyInjection.MvcServiceCollectionExtensions.AddMvc(IServiceCollection services)
   at WebApplication1.Startup.ConfigureServices(IServiceCollection services) in C:\Users\Gonkers\Development\ConsoleApp1\WebApplication1\Startup.cs:line 31
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.AspNetCore.Hosting.Internal.ConfigureServicesBuilder.Invoke(Object instance, IServiceCollection exportServices)
   at Microsoft.AspNetCore.Hosting.Internal.ConfigureServicesBuilder.<>c__DisplayClass4_0.<Build>b__0(IServiceCollection services)
   at Microsoft.AspNetCore.Hosting.ConventionBasedStartup.ConfigureServices(IServiceCollection services)
   at Microsoft.AspNetCore.Hosting.Internal.WebHost.EnsureApplicationServices()
   at Microsoft.AspNetCore.Hosting.Internal.WebHost.BuildApplication()
   at Microsoft.AspNetCore.Hosting.WebHostBuilder.Build()
   at WebApplication1.Program.Main(String[] args) in C:\Users\Gonkers\Development\ConsoleApp1\WebApplication1\Program.cs:line 16
```
I'm using corerun to launch the web app `../RavenDB.4.0/dotnet/corerun ./WebApplication1.dll`

Basic console apps seem to run just fine without any issues. 

Thanks

**Comment by pranavkm** [comment]
Could you attach the WeApplication1.deps.json file from the output directory of the application?

**Comment by Gonkers** [comment]
You can find the deps file here... https://github.com/Gonkers/Github-Issues/blob/master/Files/ConsoleApp1/WebApplication1/bin/Debug/PublishOutput/WebApplication1.deps.json

Thanks.

**Comment by pranavkm** [comment]
Looks like the issue is a result of not having a runtime component (`linux-arm` \ `linux-arm64`) for Libuv 1.9.1 on Raspberry-pi. There's a bug in the `CandidateResolver` where it expects every dependency to have a runtime component - perhaps that's not necessarily true. @pakrym does that sound like an incorrect assumption?

@moozzyk, can Kestrel 1.1.0 run on Raspberry pi 3 given that Libuv doesn't have runtime libraries targeting arm? Would using Kestrel 1.2.0 \ Libuv 1.10.0-* be a better bet here?

**Comment by ayende** [comment]
@pranavkm Kesterl most certain run on the PI, yes.
IIRC, we had to compile it there, but nothing beyond it.

**Comment by pranavkm** [comment]
@ayende - I should've worded it better - would using Libuv 1.9.1 work? Using a custom build or the 1.10.0-* build of the package that's on the nightly feeds might fix the missing runtime libraries that's showing up as part of @Gonkers's stacktrace.

**Comment by moozzyk** [comment]
The 1.10.0 fat libuv package now contains libuv for linux-arm, linux-armel and linux-arm64 (it was added last week). CoreClr is now using this package to get the runtime you don't have to build on your own (tracked here: https://github.com/dotnet/core-setup/issues/725#issuecomment-273854460). The 1.9.1 fat package did not contain the arm version of libuv so you would have to build it yourself. Also, you would probably have to build runtime yourself for arm as well.

**Comment by moozzyk** [comment]
@pranavkm - agree, I would just try to add the latest Libuv 1.10.0 to the project as a dependency (but still be careful to not bring other packages from nightly feed to avoid version hell).

**Comment by pranavkm** [comment]
Spoke to @pakrym offline - `DefaultAssemblyPartDiscoveryProvider` shouldn't expect every dependency to have a runtime component (which is essentially what this statement is asserting - https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultAssemblyPartDiscoveryProvider.cs#L103). Would be good to rename the field name to make it clear that it represents the runtime component and not the actual dependency.

**Comment by peterdeme** [comment]
Hi, I'm facing the same issue in when uploading my app to AWS Lambda. Is there any workaround?

Thanks
Peter

**Comment by pranavkm** [comment]
@peterdeme - could you include your deps file? The issue @Gonkers was running in to happened to be because there weren't native assets for an ARM package. It's unlikely you're running in to the same thing.

I'm getting a little ahead of myself, but one possible work around until we sort out your concerns would be to register an `ApplicationPartManager` in the service collection and set it up in a way that it doesn't do controller discovery using `DefaultAssemblyPartDiscoveryProvider`.

```C#
var applicationPartManager = new ApplicationPartManager();
var applicationPartAssemblies = new[] 
{ 
   typeof(Startup).GetTypeInfo().Assembly,
   typeof(MyClassLibraryWithControllers).GetTypeInfo().Assembly,
   typeof(MyCustomSetOfTagHelpers).GetTypeInfo().Assembly
};

foreach (var assembly in applicationPartAssemblies)
{
   applicationPartManager.ApplicationParts.Add(new AssemblyPart(assembly);
}
```

**Comment by peterdeme** [comment]
Hi @pranavkm 

Sure. Here it is: https://pastebin.com/tkwSCu6T

**Comment by rynowak** [comment]
I'm re-opening this for now to track the issue until we get a resolution.

**Comment by pranavkm** [comment]
@moozzyk - the Libuv package in @peterdeme 's deps file doesn't have a runtime component:

```
      "libuv/1.9.1": {
        "dependencies": {
          "Microsoft.NETCore.Platforms": "1.1.0"
        },
        "compileOnly": true
      },
```

Any idea why it might not be missing the rid specific components?

**Comment by moozzyk** [comment]
Don't know. If this repros 100% it would be good to have a repro to try it out.

**Comment by peterdeme** [comment]
Well, looks like I don't have this issue in netcore1.0. For now, we'll stick to using 1.0 for AWS Lambda. That's the deps for that 1.0 build: https://pastebin.com/VFFzWQ0R

**Comment by peterdeme** [comment]
Hi Guys,

I accidentally found an article which explains it all: http://lunarfrog.com/blog/using-net-core-for-aws-lambda-deployment

> The guide describing migration of the existing Web API site does not mention a couple of additional changes, which are required. Default .NET Core Web application template uses Microsoft.NET.Sdk.Web as a SDK. However, this SDK references libuv runtime component which is missing on AWS and raises runtime exception. This is a known issue in .NET Core and was recently fixed in v1.1 but not available in production. To avoid this issue, Web application for Lambda shall use Microsoft.NET.Sdk instead (and set EXE as an output type).

**Comment by pranavkm** [comment]
Thanks @peterdeme! I'll close this since this is a known issue with a workaround for now.
